pettingzoo/butterfly/knights_archers_zombies/knights_archers_zombies.py

```diff
@@ -261,41 +274,119 @@ def zombie_all_players(self, knight_list, archer_list, run):
-            min_y = agent_position[1] - 256
-            max_y = agent_position[1] + 256
-            lower_y_bound = max(min_y, 0)
-            upper_y_bound = min(max_y, const.SCREEN_HEIGHT)
-            lower_x_bound = max(min_x, 0)
-            upper_x_bound = min(max_x, const.SCREEN_WIDTH)
-            startx = lower_x_bound - min_x
-            starty = lower_y_bound - min_y
-            endx = 512 + upper_x_bound - max_x
-            endy = 512 + upper_y_bound - max_y
-            cropped = np.zeros_like(self.observation_spaces[agent].low)
-            cropped[startx:endx, starty:endy, :] = screen[
-                lower_x_bound:upper_x_bound, lower_y_bound:upper_y_bound, :
-            ]
-
-        return np.swapaxes(cropped, 1, 0)
+            # get the agent
+            i = self.agent_name_mapping[agent]
+            agent = self.agent_list[i]
+
+            # get the agent position
+            agent_state = agent.vector_state
+            agent_pos = np.expand_dims(agent_state[4:6], axis=0)
+            agent_ang = np.expand_dims(agent_state[6:8], axis=0)
+
+            # get vector state of everything
+            state = self.get_vector_state()
+            all_ids = state[:, 0:4]
+            all_pos = state[:, 4:6]
+            all_ang = state[:, 6:8]
+
+            # compute the angles that everything is facing
+            all_ang = np.arctan2(all_ang[:, 0], all_ang[:, 1])
+            agent_ang = np.arctan2(agent_ang[:, 0], agent_ang[:, 1])
+
+            # get relative angles that everything is facing relative
+            # to the current agent
+            rel_ang = all_ang - agent_ang
+
+            # get relative angle vector
+            c, s = np.cos(rel_ang), np.sin(rel_ang)
+            rel_ang = np.stack([c, s], axis=0).T
+
+            # get relative positions
+            rel_pos = all_pos - agent_pos
+
+            # give more emphasis to closer positions
+            rel_pos = np.sign(rel_pos) * (1 - abs(rel_pos))
+
+            # get rotation matrix of agent
+            c, s = np.cos(agent_ang), np.sin(agent_ang)
+            rot_mat = np.array([[c, -s], [s, c]])
+            rot_mat = np.squeeze(rot_mat).T
+
+            rel_pos = rel_pos @ rot_mat
+
+            # give more emphasis to closer targets
+            rel_pos = 1. / (rel_pos + 1e-6)
+
+            # combine the positions and angles
+            # set the current agents one to be absolute
+            # set the current agent's entity type to 3
+            state = np.concatenate([all_ids, rel_pos, rel_ang], axis=-1)
+            state[i] = agent_state
+            state[i, 3] = 1.
+            state = self.pad_vector_state(state)
+
+            return state
 
     def state(self):
         """
         Returns an observation of the global environment
         """
-        state = pygame.surfarray.pixels3d(self.WINDOW).copy()
-        state = np.rot90(state, k=3)
-        state = np.fliplr(state)
+        state = None
```

Comment:
Declaring variables is not necessary in python

Reply:
Fixed this, thanks! For some reason I never knew it could do that, probably cause my LSP keeps giving me unbounded errors.